### PR TITLE
damldocs: Hook up multiple file rendering to cli options.

### PR DESCRIPTION
**Pull request by associahedron**
This PR does a bit of cleanup of `DA.Daml.Doc.Driver` and hooks up the renderer backend for Rst, Markdown, Html docs. Probably should convert Hoogle docs next to work in the renderer, but I'm happy to keep Json input/output totally in the Driver as it currently is.

With this PR, rendering to a folder is the default, but the `--combine` flag will combine all the docs into a single output file.

### CODE REVIEW

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs [hunk 1]
```diff
@@ -18,67 +21,99 @@ import Development.IDE.Types.Location
...
-import           System.Directory
-import           System.FilePath
 
+data DamldocOptions = DamldocOptions
+    { do_inputFormat :: InputFormat
+    , do_ideOptions :: IdeOptions
+    , do_outputPath :: FilePath
+    , do_outputFormat :: OutputFormat
+    , do_docTemplate :: Maybe FilePath
+    , do_docOptions :: [DocOption]
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]: Having a type called `DamldocOptions` that has a list of `DocOption` seems confusing. We should probably refactor the whole handling of `DocOption` at some point. But that’s probably better left for a separate PR.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs [hunk 2]
```diff
@@ -18,67 +21,99 @@ import Development.IDE.Types.Location
...
-              -> DocFormat
-              -> Maybe FilePath
-              -> [DocOption]
-              -> [NormalizedFilePath]
-              -> IO ()
-damlDocDriver cInputFormat ideOpts output cFormat prefixFile options files = do
+data OutputFormat = OutputJson | OutputHoogle | OutputDocs RenderFormat
+    deriving (Eq, Show, Read)
+
+-- | Run damldocs!
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]: Maybe calle it `runDamlDoc`s then? Not a big fan of calling things `driver`.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs [hunk 3]
```diff
@@ -18,67 +21,99 @@ import Development.IDE.Types.Location
...
-                applyAnnotations <$> concatMapM (either printAndExit pure) mbData
+        InputDaml -> onErrorExit . runExceptT $
+            fmap (applyTransform do_docOptions)
+                (mkDocs do_ideOptions do_inputFiles)
 
-            InputDaml ->
-                onErrorExit $ runExceptT
-                            $ fmap (applyTransform options)
-                            $ mkDocs ideOpts files
+-- | Transform doc data, applying annotations and
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]: Looks like that comment stops in the middle of a sentence. Given that this only seems to be used once and I don’t find either the name or the comment to add much helpful information, I would just inline the implementation.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs [hunk 4]
```diff
@@ -18,67 +21,99 @@ import Development.IDE.Types.Location
...
+transformDocData :: DamldocOptions -> [ModuleDoc] -> [ModuleDoc]
+transformDocData DamldocOptions {..} = applyTransform do_docOptions
 
-    prefix <- maybe (pure "") BS.readFile prefixFile
+-- | Output doc data.
+renderDocData :: DamldocOptions -> [ModuleDoc] -> IO ()
+renderDocData DamldocOptions{..} docData = do
+    templateM <- mapM (fmap T.decodeUtf8 . BS.readFile) do_docTemplate
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]:
```suggestion
    templateM <- mapM T.readFileUtf8 do_docTemplate
```
with the helper from `Data.Text.Extended`.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs [hunk 5]
```diff
@@ -18,67 +21,99 @@ import Development.IDE.Types.Location
...
             putStrLn $ "Writing " ++ file ++ "..."
-            BS.writeFile file $ prefix <> BS.fromStrict (T.encodeUtf8 contents)
-
-    case cFormat of
-            Json -> write output $ T.decodeUtf8 . BS.toStrict $ AP.encodePretty' jsonConf docData
-            Rst  -> write output . renderPage . mconcat $ map renderSimpleRst docData
-            Hoogle   -> write output . T.concat $ map renderSimpleHoogle docData
-            Markdown -> write output . renderPage . mconcat $ map renderSimpleMD docData
-            Html -> do
+            BS.writeFile file . T.encodeUtf8 $ prefix <> contents
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]:
```suggestion
            T.writeFileUtf8 $ prefix <> contents
```

File: compiler/damlc/lib/DA/Cli/Damlc/Command/Damldoc.hs [hunk 6]
```diff
@@ -62,33 +67,44 @@ documentation = Damldoc
...
+            metavar "FORMAT"
+            <> help "Output format. Valid format names: rst, md, markdown, html, hoogle, json (Default: markdown)."
+            <> short 'f'
+            <> long "format"
+            <> value (OutputDocs Markdown)
+
+    readOutputFormat =
+        eitherReader $ \arg ->
+            case lower arg of
+                "rst" -> Right (OutputDocs Rst)
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Command/Damldoc.hs]: Thanks! I kept getting the format names wrong :slightly_smiling_face:

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs (hunk 1 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]: I agree :-)

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs (hunk 3 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]: Fair

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs (hunk 4 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]: thanks!

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs (hunk 2 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs]: 👍

File: compiler/damlc/lib/DA/Cli/Damlc/Command/Damldoc.hs (hunk 6 above)
Comment (associahedron) [reply compiler/damlc/lib/DA/Cli/Damlc/Command/Damldoc.hs]: :-)

**Comment by jberthold-da** [comment]
:+1:  prettier. Two points to follow up:
* Might be good to add to the changelog?
* I think `--json` should have a different behaviour, namely to switch input format to json, or not exist at all. Mentioned earlier, but we are still carrying it around with the prior semantics

**Comment by associahedron** [comment]
> 👍 prettier. Two points to follow up:
> 
> * Might be good to add to the changelog?
> * I think `--json` should have a different behaviour, namely to switch input format to json, or not exist at all. Mentioned earlier, but we are still carrying it around with the prior semantics

Thanks :) I'm happy to drop `--json`.

**Pull request merged**